src/signature.c

```diff
@@ -0,0 +1,178 @@
+#include <errno.h>
+#include <fcntl.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/uio.h>
+#include <unistd.h>
+
+#include "memory.h"
+#include "process.h"
+#include "signature.h"
+
+#include <luajit.h>
+
+extern game_process process;
+
+MemoryRegion* get_memory_regions(pid_t pid, int* count)
+{
+    char maps_path[256];
+    sprintf(maps_path, "/proc/%d/maps", pid);
+    FILE* maps_file = fopen(maps_path, "r");
+    if (!maps_file) {
+        perror("Failed to open maps file");
+        return NULL;
+    }
+
+    MemoryRegion* regions = NULL;
+    int capacity = 0;
+    *count = 0;
+
+    char line[256];
+    while (fgets(line, sizeof(line), maps_file)) {
+        if (*count >= capacity) {
+            capacity = capacity == 0 ? 10 : capacity * 2;
+            regions = (MemoryRegion*)realloc(regions, capacity * sizeof(MemoryRegion));
+            if (!regions) {
+                perror("Failed to allocate memory for regions");
+                fclose(maps_file);
+                return NULL;
+            }
+        }
+
+        uintptr_t start, end;
+        if (sscanf(line, "%lx-%lx", &start, &end) != 2) {
+            continue; // Skip lines that don't match the expected format
+        }
+        regions[*count].start = start;
+        regions[*count].end = end;
+        (*count)++;
+    }
+
+    fclose(maps_file);
+    return regions;
+}
+
+bool match_pattern(const uint8_t* data, const int* pattern, size_t pattern_size)
+{
+    for (size_t i = 0; i < pattern_size; ++i) {
+        if (pattern[i] != -1 && data[i] != pattern[i])
+            return false;
+    }
+    return true;
+}
+
+int* convert_signature(const char* signature, size_t* pattern_size)
+{
+    char* signature_copy = strdup(signature);
+    if (!signature_copy) {
+        return NULL;
+    }
+
+    char* token = strtok(signature_copy, " ");
+    size_t size = 0;
+    size_t capacity = 10;
+    int* pattern = (int*)malloc(capacity * sizeof(int));
+    if (!pattern) {
+        free(signature_copy);
+        return NULL;
+    }
+
+    while (token != NULL) {
+        if (size >= capacity) {
+            capacity *= 2;
+            int* temp = (int*)realloc(pattern, capacity * sizeof(int));
+            if (!temp) {
+                free(pattern);
+                free(signature_copy);
+                return NULL;
+            }
+            pattern = temp;
+        }
+
+        if (strcmp(token, "??") == 0)
+            pattern[size] = -1;
```

Comment:
Setting -1 as the unknown marker on an uint8_t is the same as setting it to 255 (0xFF), so if the user asks for a pattern that has 0xFF it will completely ignore that byte. And setting it to int8_t is not really a good option either, so the only way would be to use an uint16_t and only check for the lower byte and if the higher byte is 1 then ignore it, at least thats my idea, but it would require a lot more changes and would also make things way more complicated than they already are
![image](https://github.com/user-attachments/assets/31b21944-1283-47c9-80e7-8440ca7a7d23)

Reply:
I think I've fixed this in my latest commit? I ran with your idea of turning it into a uint16_t. An FF byte in the string should now be converted to `0x00FF` and an FF byte in the string should be converted to `0xFF00`.